scripts/bundles/plugins/typescript-source-plugin.ts

```diff
@@ -52,67 +52,67 @@ async function bundleTypeScriptSource(tsPath: string, opts: BuildOptions): Promi
-  code = removeFromSource(code, `ts.getDefaultLibFilePath = getDefaultLibFilePath;`);
+  code = removeFromSource(code, `getDefaultLibFilePath: () => getDefaultLibFilePath,`);
 
   // remove the CPUProfiler since it uses node apis
-  code = removeFromSource(code, `enableCPUProfiler: enableCPUProfiler,`);
-  code = removeFromSource(code, `disableCPUProfiler: disableCPUProfiler,`);
+  code = removeFromSource(code, `enableCPUProfiler,`);
+  code = removeFromSource(code, `disableCPUProfiler,`);
+
+  // this is a ~~sketchy~~ thing to do!
+  const TS_PROP_DEFINER = `__defProp(target, name, { get: all[name], enumerable: true });`;
+  const TS_PROP_DEFINER_DRASTICALLY_IMPROVED = `__defProp(target, name, { get: all[name], enumerable: true, configurable: true });`;
+
+  code = code.replace(TS_PROP_DEFINER, TS_PROP_DEFINER_DRASTICALLY_IMPROVED);
+
+  // this is the bit where we transmute a commonjs module into an ES module
+  // don't ask why, just go with it 🙃
 
   // trim off the last part that sets module.exports and polyfills globalThis since
   // we don't want typescript to add itself to module.exports when in a node env
-  const tsEnding = `})(ts || (ts = {}));`;
+  const tsEnding = `if (typeof module !== "undefined" && module.exports) { module.exports = ts; }`;
+
   if (!code.includes(tsEnding)) {
     throw new Error(`"${tsEnding}" not found`);
   }
   const lastEnding = code.lastIndexOf(tsEnding);
-  code = code.slice(0, lastEnding + tsEnding.length);
+  code = code.slice(0, lastEnding);
 
-  // there's a billion unnecessary "var ts;" for namespaces
-  // but we'll be using the top level "const ts" instead
-  code = code.replace(/var ts;/g, '');
-
-  // minification is crazy better if it doesn't use typescript's
-  // namespace closures, like (function(ts) {...})(ts = ts || {});
-  code = code.replace(/ \|\| \(ts \= \{\}\)/g, '');
-
-  // make a nice clean default export
-  // "process.browser" is used by typescript to know if it should use the node sys or not
   const o: string[] = [];
   o.push(`// TypeScript ${opts.typescriptVersion}`);
   o.push(`import { IS_NODE_ENV } from '@environment';`);
   o.push(`process.browser = !IS_NODE_ENV;`);
-  o.push(`const ts = {};`);
   o.push(code);
   o.push(`export default ts;`);
   code = o.join('\n');
 
   const { minify } = await import('terser');
 
-  if (opts.isProd) {
-    const minified = await minify(code, {
-      ecma: 2018,
-      module: true,
-      compress: {
-        ecma: 2018,
-        passes: 2,
-      },
-      format: {
-        ecma: 2018,
-        comments: false,
-      },
-    });
-    code = minified.code;
-  }
+  // if (opts.isProd) {
+  //   const minified = await minify(code, {
+  //     ecma: 2018,
+  //     module: true,
+  //     compress: {
+  //       ecma: 2018,
+  //       passes: 2,
+  //     },
+  //     format: {
+  //       ecma: 2018,
+  //       comments: false,
+  //     },
+  //   });
+  //   code = minified.code;
+  // }
```

Comment:
I ran into some problems with some side effect of the minification here that I haven't yet been able to track down.

I question whether this is necessary however.

This will basically minify `lib/typescript.js` _before_ it gets incorporated into the `compiler` bundle which uses this plugin. However, we then later produce both minified and non-minified output from the compiler. Something like:

```mermaid
flowchart LR
    A[minified typescript] -->|included in| B{Compiler bundle}
    B --> C[minified output]
    B --> D[non-minified output]
```

My question is, if we're already minifying downstream, do we need to do the minification of just the TS source as well? Could we instead just:

```mermaid
flowchart LR
    A[typescript] -->|included in| B{Compiler bundle}
    B --> C[minified output]
    B --> D[non-minified output]
```

and defer the minification to a single pass, when we build the compiler bundle?

One advantage of this would be non-minified typescript in the compiler bundle, which would I think make it easier to track down issues if they occur somewhere in typescript-land. At present this is a big unfortunate minified blob :/

Reply:
A little data I just gathered on this

After building w/ typescript five, building from this branch, without the initial minification done here in the `typescriptSourcePlugin`:

- .tgz size is 8.3mb
- `compiler/stencil.js` is 10mb
- `compiler/stencil.min.js` is 4.3mb

building w/ typescript five, with minification here in `typescriptSourcePlugin`:

- .tgz size is 6.6mb
- `compiler/stencil.js` is 5.7mb
- `compiler/stencil.min.js` is 4.3mb

on `main` (so with 4.9):

- .tgz size is 6.8mb
- `compiler/stencil.js` is 6.1mb
- `compiler/stencil.min.js` is 4.6mb

So to summarize:

- it looks like typescript 5 is overall smaller, since when minification is turned on it results in smaller files than on typescript 4.9. This makes sense given what the typescript team said about switching to esbuild.
- skipping the first pass of minification has a fairly sizable penalty on bundle size (2.6mb bigger compared to 5.7mb is a ~45% increase)
- likewise, the size of `compiler/stencil.js` grows significantly without minifying typescript

I think I'll make some efforts today to get the minification working correctly with typescript 5 so we have the option if we want to keep it.

Reply:
Alright so I did some more investigation yesterday afternoon and was able to come to some conclusions about _where_ the problem is happening, but as of yet I don't have a solution for getting the minification to work w/ typescript 5.

The problem is occurring in the `createTsBuildProgram`, in this call to `ts.createWatchProgram`:

https://github.com/ionic-team/stencil/blob/ae23e929504e27d4b5e80a68031f6301803c1292/src/compiler/transpile/create-build-program.ts#L124-L124

by adding a bunch of debug statements to the typescript source I figured out that the problem is occurring on this line:

[src/compiler/watchPublic.ts:L529-L531](https://github.com/microsoft/TypeScript/blob/58f030072ed41e3428c658a83579bdb880ef6886/src/compiler/watchPublic.ts#L529-L531)
```ts
    const customHasInvalidatedResolutions = userProvidedResolution ?
        maybeBind(host, host.hasInvalidatedResolutions) || returnTrue :
        returnFalse;
```

but...idk why haha. Something is not working right when this is minified, but working fine when it's _not_ minified.

Since I haven't found a good solution, my inclination at this point is to file a tech debt issue to investigate re-enabling minification in the future and go forward without it for now. @rwaskiewicz thoughts?